core/libp2p/libp2p.hpp

```diff
@@ -0,0 +1,127 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_LIBP2P_HPP
+#define KAGOME_LIBP2P_HPP
+
+#include <memory>
+
+#include <rxcpp/rx.hpp>
+#include "common/result.hpp"
+#include "libp2p/common_objects/multiaddress.hpp"
+#include "libp2p/common_objects/multihash.hpp"
+#include "libp2p/common_objects/peer.hpp"
+#include "libp2p/connection/connection.hpp"
+#include "libp2p/connection/muxed_connection.hpp"
+#include "libp2p/discovery/peer_discovery.hpp"
+#include "libp2p/muxer/muxer.hpp"
+#include "libp2p/routing/router.hpp"
+#include "libp2p/store/record.hpp"
+#include "libp2p/store/record_store.hpp"
+#include "libp2p/swarm/dial_status.hpp"
+#include "libp2p/swarm/swarm.hpp"
+#include "libp2p/transport/transport.hpp"
+#include "libp2p/transport/transport_listener.hpp"
+
+namespace libp2p {
+  /**
+   * Provides a way to instantiate a libp2p instance via joining different
+   * components to it and dial to the peer
+   */
+  class Libp2p {
+    /**
+     * Add a swarm component
+     * @param swarm to be added
+     */
+    virtual void addSwarm(std::shared_ptr<swarm::Swarm> swarm) = 0;
```

Comment:
why do you accept shared ptr as argument? what is the purpose of this method?

Reply:
This is an interface-recommended way to instantiate libp2p instance. Shared_ptr, because I assume those components to be created somewhere and then passed here. On the other hand, they also can unique_ptr-s, as maybe they aren't going to be used anywhere else